src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs

```diff
@@ -163,12 +179,19 @@ protected virtual bool ShouldOutputNuGetVersion
             get { return Console.Verbosity == Verbosity.Detailed; }
         }
 
+        protected virtual void SetDefaultCredentialProvider()
+        {
+            SetDefaultCredentialProvider(MsBuildDirectory);
+        }
+
         /// <summary>
         /// Set default credential provider for the HttpClient, which is used by V2 sources.
         /// Also set up authenticated proxy handling for V3 sources.
         /// </summary>
-        protected void SetDefaultCredentialProvider()
+        protected void SetDefaultCredentialProvider(Lazy<string> msbuildDirectory)
```

Comment:
wouldn't it make more sense for the function signature to just accept a string, and the caller pass MsBuildDirectory.Value to the function call?

Reply:
This becomes more important later in https://github.com/NuGet/NuGet.Client/pull/2208/files

We only want to discover MSBuild when we absolutely have to. 
That means we only do it when we attempt to use the plugins.

Reply:
that's fine. But given that calling this function essentially means that msbuild is going to be discovered for sure, shouldn't it be the responsibility of the caller to do the discovery and pass the path directly to this function?

Reply:
This call is made for every command. 

MSBuild needs to discovered only when we have restore or we're using http. 
This approach ensures that.